src/connector/src/source/kafka/source/mux_reader.rs

```diff
@@ -0,0 +1,288 @@
+    /// (topic, partition) -> sender (each topic unique within this connection)
+    senders: RwLock<HashMap<PartitionRoute, mpsc::Sender<Vec<OwnedMessage>>>>,
+}
+
+static GLOBAL: OnceCell<RwLock<HashMap<ReaderKey, Arc<KafkaMuxReader>>>> = OnceCell::new();
+
+impl KafkaMuxReader {
+    fn registry() -> &'static RwLock<HashMap<ReaderKey, Arc<KafkaMuxReader>>> {
+        GLOBAL.get_or_init(|| RwLock::new(HashMap::new()))
+    }
+
+    /// Create or reuse the reader for a given connection.
+    pub async fn get_or_create(
+        connection_id: ReaderKey,
+        properties: KafkaProperties,
+        source_ctx: SourceContextRef,
+    ) -> Result<Arc<Self>> {
+        // fast path – already exists
+        if let Some(r) = Self::registry().read().await.get(&connection_id).cloned() {
+            tracing::info!("Reusing existing KafkaMuxReader for connection_id: {connection_id}");
+            return Ok(r);
+        }
+
+        let mut config = ClientConfig::new();
+
+        let bootstrap_servers = &properties.connection.brokers;
+        let broker_rewrite_map = properties.privatelink_common.broker_rewrite_map.clone();
+
+        // disable partition eof
+        config.set("enable.partition.eof", "false");
+        config.set("auto.offset.reset", "smallest");
+        config.set("isolation.level", KAFKA_ISOLATION_LEVEL);
+        config.set("bootstrap.servers", bootstrap_servers);
+
+        properties.connection.set_security_properties(&mut config);
+        properties.set_client(&mut config);
+
+        let prefix = properties
+            .group_id_prefix
+            .as_deref()
+            .unwrap_or("rw-consumer");
+
+        config.set("group.id", format!("{prefix}-mux-{connection_id}"));
+
+        let ctx_common = KafkaContextCommon::new(
+            broker_rewrite_map,
+            Some(format!("shared-mux-reader-connection-{}", connection_id,)),
+            // thread consumer will keep polling in the background, we don't need to call `poll`
+            // explicitly
+            Some(source_ctx.metrics.rdkafka_native_metric.clone()),
+            properties.aws_auth_props,
+            properties.connection.is_aws_msk_iam(),
+        )
+        .await?;
+
+        let client_ctx = RwConsumerContext::new(ctx_common);
+
+        if let Some(log_level) = read_kafka_log_level() {
+            config.set_log_level(log_level);
+        }
+        let consumer: StreamConsumer<RwConsumerContext> = config
+            .create_with_context(client_ctx)
+            .await
+            .context("failed to create kafka consumer")?;
+
+        let reader = Arc::new(Self {
+            key: connection_id.clone(),
+            consumer,
+            senders: RwLock::new(HashMap::new()),
+        });
+
+        let max_chunk_size = source_ctx.source_ctrl_opts.chunk_size;
+
+        Self::registry()
+            .write()
+            .await
+            .insert(connection_id, Arc::clone(&reader));
```

Comment:
This doesn't seems to ensure single-flight. Shall we consider `moka`?

Reply:
+1